### PR TITLE
upgrade ndarray to 0.16

### PR DESCRIPTION
**Pull request by nmandery**
This PR upgrades the ndarray dependency from 0.15 to 0.16

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

### CODE REVIEW

**Comment by lnicola** [comment]
Thanks, but I don't think this is right (just as the previous code wasn't right). For reference, I saw this example in https://github.com/rust-ndarray/ndarray/pull/994:

```rust
let a = Array2::<f64>::zeros([3, 4]).slice_move(s![1.., ..]);
let b = Array2::<f64>::zeros([2, 4]);
assert_eq!(a.shape(), b.shape());
assert_eq!(a.strides(), b.strides());
assert_ne!(
    unsafe { a.as_ptr().offset_from(a.into_raw_vec().as_ptr()) },
    unsafe { b.as_ptr().offset_from(b.into_raw_vec().as_ptr()) },
);
```

Which implies that slicing sometimes avoids copying the data by bumping an offset from the start of the storage (the example strips off the first row).

So the `offset` is not always 0 (of course, otherwise it wouldn't be there), and we should handle it, either by copying the data to a new `Vec`, or by introducing a similar `offset` concept, which would probably be quite confusing to use.

**Comment by nmandery** [comment]
I understood having a standard layout means also having an offset of 0, but I am afraid you are right. So we need to copy the data (like it is currently done for the non-standard-layout) in case of an offset.

I will add this to the PR later today.

**Comment by lnicola** [comment]
I'm again wondering if we should fully switch to `ndarray` (see #570).

File: src/raster/buffer.rs [hunk 1]
```diff
@@ -220,7 +220,12 @@ impl<T: GdalType + Copy> From<Array2<T>> for Buffer<T> {
         let shape = value.shape();
         let (cols, rows) = (shape[1], shape[0]);
         let data: Vec<T> = if value.is_standard_layout() {
-            value.into_raw_vec()
+            let (data, offset) = value.into_raw_vec_and_offset();
+            if let Some(offset) = offset {
+                data.into_iter().skip(offset).collect()
+            } else {
+                data
```
Comment (lnicola) [inline src/raster/buffer.rs:227]: Let's also handle `offset == 0` the same way.

Comment (nmandery) [reply src/raster/buffer.rs:227]: Done - I removed the if

File: src/raster/buffer.rs [hunk 2]
```diff
@@ -220,7 +220,12 @@ impl<T: GdalType + Copy> From<Array2<T>> for Buffer<T> {
         let shape = value.shape();
         let (cols, rows) = (shape[1], shape[0]);
         let data: Vec<T> = if value.is_standard_layout() {
-            value.into_raw_vec()
+            let (data, offset) = value.into_raw_vec_and_offset();
+            if let Some(offset) = offset {
+                data.into_iter().skip(offset).collect()
```
[the remark this answers is not in the document]
Comment (nmandery) [reply src/raster/buffer.rs]: Yepp, done

File: src/raster/buffer.rs (hunk 1 above)
Comment (lnicola) [reply src/raster/buffer.rs]: That might do an extra copy in the zero offset case.

**Comment by lnicola** [comment]
@rouault looks like I've just won the lottery in https://github.com/georust/gdal/actions/runs/11372075639/job/31635584318:

> OGR2SQLITE_Setup() failed due to sqlite3_api == nullptr

**Pull request merged**